libs/sysplugins/smarty_internal_runtime_writefile.php

```diff
@@ -89,11 +83,8 @@ public function writeFile($_filepath, $_contents, Smarty $smarty)
             error_reporting($_error_reporting);
             throw new SmartyException("unable to write file {$_filepath}");
         }
-        if ($_file_perms !== null) {
-            // set file permissions
-            chmod($_filepath, $_file_perms);
-            umask($old_umask);
-        }
+        // set file permissions
+        chmod($_filepath, 0666 & ~umask());
```

Comment:
Fixed chmod mode to apply the umask correctly

Reply:
If I were you, I'd simply remove all calls to chmod()/chown(). This is not a file management code, and unless you do something drastic, you should be able to read files you've created. Meaning, no chmod required.

If there's still a problem, then fix your FS settings. All modern FS supports ACL, you can get results much better than ancient POSIX permissions.

Reply:
Again, both your cases fall into "fix your FS permissions already" category.

```
setfacl -Rm "d:u:builder-user:rwX,d:u:fpm-user:rwX,d:m::rwx" -- /templates/directory
```

Reply:
@AnrDaemon,

setfacl is one of the ways to solve the issue with permissions.

I suggest to improve work with chmod and add the ability to influence on chmod which is hardcoded

Reply:
@AnrDaemon
Please also note that it is possible to set the default umask for php-fpm, for instance.
I usually do that and set umask to 0027.

smarty will nonetheless always create cache files with 0644.
Having explicitly set a umask, my expectation is that this default is honored and used.

It is certainly true that you can fine-tune things using ACLs in ways file permissions don't allow for. However, dismissing POSIX as ancient and thus something that can be ignored is in and of itself a rather disturbing point of view, to be frank.